pygfx/geometries/_text.py

```diff
@@ -576,65 +630,208 @@ def _apply_layout(self):
+                )
             i1, i2 = item.offset, item.offset + positions.shape[0]
             positions_array[i1:i2] = positions
+            # Keep a pointer to the array so we can align the text
+            current_line.append(positions_array[i1:i2])
+
             sizes_array[i1:i2] = font_size
+
             # Prepare for next
+            extent_offset += item.extent * font_size
             ws_margin = item.margin_after * font_size
-            extent_offset += item.extent * font_size + ws_margin
-            # Update total extent
+            extent_offset += ws_margin
+
+            # Update line extent
             if is_horizontal:
-                right = extent_offset
-                top = max(top, item.ascender * font_size)
-                bottom = min(bottom, item.descender * font_size)
+                line_right = extent_offset
+                line_top = max(line_top, item.ascender * font_size)
+                line_bottom = min(
+                    line_bottom, item.descender * font_size + vertical_offset
+                )
             else:
-                bottom = extent_offset
-                right = max(right, item.ascender * font_size)
-                left = min(left, item.descender * font_size)
+                line_bottom = extent_offset
+                line_right = max(line_right, item.ascender * font_size)
+                line_left = min(line_left, item.descender * font_size + vertical_offset)
+
+            # Update total extent
+            top = max(top, line_top)
+            bottom = min(bottom, line_bottom)
+            right = max(right, line_right)
+            left = min(left, line_left)
+
+            if item.newline_after:
+                lines.append(current_line)
+                lines_aabb.append(
+                    np.array(
+                        [(line_left, line_bottom, 0), (line_right, line_top, 0)],
+                        np.float32,
+                    )
+                )
+                current_line = []
+                line_left = line_right = 0
+                line_top = line_bottom = 0
+                vertical_offset -= len(item.newline_after) * line_height
+                extent_offset = 0
+
+        lines.append(current_line)
+        lines_aabb.append(
+            np.array(
+                [(line_left, line_bottom, 0), (line_right, line_top, 0)], np.float32
+            )
+        )
+        current_line = []
+        line_left = line_right = 0
+        line_top = line_bottom = 0
+
+        # take care of new lines at the end of the text
+        bottom = min(bottom, vertical_offset)
 
         self._aabb = np.array([(left, bottom, 0), (right, top, 0)], np.float32)
 
         # Anchoring
 
         if anchor.endswith("left"):
-            pos_offset_x = 0
+            pos_offset_x = -left
         elif anchor.endswith("center"):
-            pos_offset_x = -0.5 * (right - left)
+            pos_offset_x = -0.5 * (left + right)
```

Comment:
This fixes: https://github.com/pygfx/pygfx/issues/727

Reply:
Unfortunately there are other bugs with new lines in ttb text orientation...

Again, not sure if we can cut the scope here, but yeah, i tried....

You can see them with the feature example setting the direction to ttb